### PR TITLE
Collada texture images

### PR DESCRIPTION
**Pull request by Son-HNguyen**
##### Additions

* The COLLADA/KML/glTF export now supports exporting textures to binary glTF files `.glb`.

* The COLLADA/KML/glTF export now supports exporting textures of `.gltf` files in Draco compression.

##### Changes

* The option to embed textures in `.gltf` files is now set to `true` by default to reduce network calls while visualizing. If not selected, the texture images shall be moved to the folder `Appearances` to reduce redundancies.

* The option to embed textures will be automatically activated once the option to export in `.glb` is selected.

* Optimized COLLADA/KML/glTF export to prevent redundant texture as well as visualization files of implicit geometries from being exported repeatedly leading to file redundancy. This includes the following changes:

  + Exported textures are now stored in the folder `Appearances`.  Each texture has a unique name, i.e. no texture shall be overwritten.
  
  + COLLADA `.dae` and `.gltf` files of implicit geometries shall be exported in the folder `ImplicitObjects`. Similarly, they all have unique name to prevent overwriting. The `.kml` file in each tile e.g. `X/Y` shall point to these files, so that the 3DCityDB-Web-Map-Client can visualize them as usual. Empty folders such as `X/Y/<gmlid>` that used to contain the now relocated `.dae` and `.gltf` files shall be removed.

Screenshot from 2020-03-03 11-32-53

### CODE REVIEW

**Comment by yaozhihang** [comment]
I've made some tests with Exporter and Webclient. looks good!
Two remarks: 
1. Maybe the folder `ImplicitGeometry` should be renamed to e.g. `ImplicitObject(s)` or `LibraryObject(s)`, because the dae and gltf objects also contain non-geometry information. 
2. Wenn activating the Collada/glTF display rendering option `Crop texture images`, a number of non-used empty JPEG images are created. Maybe a tiny bug.

**Comment by yaozhihang** [comment]
maybe it is better to rename the folder name "appearance" to "Apperances", so that the folder names "Appearances", "Tiles", and "ImplicitObjects" have a consistent form?

**Comment by clausnagel** [comment]
What is the status of this PR?

**Comment by yaozhihang** [comment]
Son still has to look into an issue where the textureAtlas export is not correct. Once the issue has been fixed by Son, I'll resolve the git branch conflicts, and merge the PR.

**Comment by Son-HNguyen** [comment]
The fix for texture atlas in implicit objects has been committed. I did some tests using the railways dataset and so far so good. @yaozhihang please review.

**Comment by yaozhihang** [comment]
@Son-HNguyen thanks for the fix. The export of texture atlas works well now.
However, there is a further bug related to the export of implicit objects. When you export the vegetation objects in KMZ format, the Google Earth throws the following error message while loading the KMZ data:

 _"..../ImplicitObject/38.dae" Fehler: Datei konnte nicht gelesen werden_

**Comment by Son-HNguyen** [comment]
@yaozhihang Thanks for the info. I'm working on it.

File: impexp-client/src/main/java/org/citydb/modules/kml/gui/preferences/GeneralPanel.java [hunk 1]
```diff
@@ -429,7 +458,10 @@ private void setEnabledComponents() {
 		gltfConverterBrowseText.setEnabled(createGltfCheckbox.isSelected());
 		gltfConverterBrowseButton.setEnabled(createGltfCheckbox.isSelected());			
 		notCreateColladaCheckbox.setEnabled(createGltfCheckbox.isSelected());
-		embedTexturesInGltfCheckbox.setEnabled(createGltfCheckbox.isSelected());
+		// if glb is enabled, always embed textures
+		embedTexturesInGltfCheckbox.setEnabled(createGltfCheckbox.isSelected() && !exportGltfBinary.isSelected());
```
Comment (Son-HNguyen) [inline impexp-client/src/main/java/org/citydb/modules/kml/gui/preferences/GeneralPanel.java:462]: This line is not included in the branch glb-draco-support.

File: impexp-client/src/main/java/org/citydb/modules/kml/gui/preferences/GeneralPanel.java [hunk 2]
```diff
@@ -317,6 +327,19 @@ public void actionPerformed(ActionEvent e) {
...
+			@Override
+			public void itemStateChanged(ItemEvent e) {
+				if (exportGltfBinary.isSelected()) {
+					embedTexturesInGltfCheckbox.setEnabled(false);
+					embedTexturesInGltfCheckbox.setSelected(true);
+				} else {
+					embedTexturesInGltfCheckbox.setEnabled(true);
+				}
+			}
+		});
```
Comment (Son-HNguyen) [inline impexp-client/src/main/java/org/citydb/modules/kml/gui/preferences/GeneralPanel.java:342]: This part is not included in the branch glb-draco-support.

**Comment by Son-HNguyen** [comment]
This branch shall be split into two smaller branches for better management. The first branch is [glb-draco-support](https://github.com/3dcitydb/importer-exporter/tree/glb-draco-support) containing the support for export of glb and with Draco compression. The second branch will contain the changes to how we deal with implicit objects.

File: CHANGES.md [hunk 3]
```diff
@@ -48,6 +55,28 @@ interior start and end points of the segments were imported into the database. T
...
 
+##### Additions
+
+* The COLLADA/KML/glTF export now supports exporting textures to binary glTF files `.glb`.
+
+* The COLLADA/KML/glTF export now supports exporting textures of `.gltf` files in Draco compression.
+
+##### Changes
+
+* The option to embed textures in `.gltf` files is now set to `true` by default to reduce network calls while visualizing.
```
Comment (Son-HNguyen) [inline CHANGES.md:66]: Moved to branch [glb-draco-support](https://github.com/3dcitydb/importer-exporter/tree/glb-draco-support).

File: impexp-client/src/main/java/org/citydb/modules/kml/gui/preferences/GeneralPanel.java [hunk 4]
```diff
@@ -79,6 +81,8 @@
 	private JButton gltfConverterBrowseButton = new JButton("");
 	private JCheckBox notCreateColladaCheckbox = new JCheckBox();
 	private JCheckBox embedTexturesInGltfCheckbox = new JCheckBox();
+	private JCheckBox exportGltfBinary = new JCheckBox();
+	private JCheckBox enableGltfDracoCompression = new JCheckBox();
```
Comment (Son-HNguyen) [inline impexp-client/src/main/java/org/citydb/modules/kml/gui/preferences/GeneralPanel.java:85]: Moved to branch [glb-draco-support](https://github.com/3dcitydb/importer-exporter/tree/glb-draco-support).

File: impexp-client/src/main/java/org/citydb/modules/kml/gui/preferences/GeneralPanel.java [hunk 5]
```diff
@@ -135,6 +139,8 @@ public boolean isModified() {
 		if (!gltfConverterBrowseText.getText().equals(kmlExporter.getPathOfGltfConverter())) return true;
 		if (notCreateColladaCheckbox.isSelected() != kmlExporter.isNotCreateColladaFiles()) return true;
 		if (embedTexturesInGltfCheckbox.isSelected() != kmlExporter.isEmbedTexturesInGltfFiles()) return true;
+		if (exportGltfBinary.isSelected() != kmlExporter.isExportGltfBinary()) return true;
+		if (enableGltfDracoCompression.isSelected() != kmlExporter.isEnableGltfDracoCompression()) return true;
```
Comment (Son-HNguyen) [inline impexp-client/src/main/java/org/citydb/modules/kml/gui/preferences/GeneralPanel.java:143]: Moved to branch [glb-draco-support](https://github.com/3dcitydb/importer-exporter/tree/glb-draco-support).

File: impexp-client/src/main/java/org/citydb/modules/kml/gui/preferences/GeneralPanel.java [hunk 6]
```diff
@@ -161,6 +167,8 @@ private void initGui(ViewController viewController) {
 		createGltfCheckbox.setIconTextGap(10);
 		notCreateColladaCheckbox.setIconTextGap(10);
 		embedTexturesInGltfCheckbox.setIconTextGap(10);
+		exportGltfBinary.setIconTextGap(10);
+		enableGltfDracoCompression.setIconTextGap(10);
```
Comment (Son-HNguyen) [inline impexp-client/src/main/java/org/citydb/modules/kml/gui/preferences/GeneralPanel.java:171]: Moved to branch [glb-draco-support](https://github.com/3dcitydb/importer-exporter/tree/glb-draco-support).

File: impexp-client/src/main/java/org/citydb/modules/kml/gui/preferences/GeneralPanel.java [hunk 7]
```diff
@@ -169,8 +177,10 @@ private void initGui(ViewController viewController) {
...
 		collada2gltfConverterPanel.add(embedTexturesInGltfCheckbox, GuiUtil.setConstraints(0,3,1.0,1.0,GridBagConstraints.BOTH,0,BORDER_THICKNESS*5,0,BORDER_THICKNESS));
-		collada2gltfConverterPanel.add(exportGltfV1, GuiUtil.setConstraints(0,4,1.0,1.0,GridBagConstraints.BOTH,0,BORDER_THICKNESS*5,0,BORDER_THICKNESS));
-		collada2gltfConverterPanel.add(exportGltfV2, GuiUtil.setConstraints(0,5,1.0,1.0,GridBagConstraints.BOTH,0,BORDER_THICKNESS*5,0,BORDER_THICKNESS));
+		collada2gltfConverterPanel.add(exportGltfBinary, GuiUtil.setConstraints(0,4,1.0,1.0,GridBagConstraints.BOTH,0,BORDER_THICKNESS*5,0,BORDER_THICKNESS));
+		collada2gltfConverterPanel.add(enableGltfDracoCompression, GuiUtil.setConstraints(0,5,1.0,1.0,GridBagConstraints.BOTH,0,BORDER_THICKNESS*5,0,BORDER_THICKNESS));
+		collada2gltfConverterPanel.add(exportGltfV1, GuiUtil.setConstraints(0,6,1.0,1.0,GridBagConstraints.BOTH,0,BORDER_THICKNESS*5,0,BORDER_THICKNESS));
+		collada2gltfConverterPanel.add(exportGltfV2, GuiUtil.setConstraints(0,7,1.0,1.0,GridBagConstraints.BOTH,0,BORDER_THICKNESS*5,0,BORDER_THICKNESS));
```
Comment (Son-HNguyen) [inline impexp-client/src/main/java/org/citydb/modules/kml/gui/preferences/GeneralPanel.java:183]: Moved to branch [glb-draco-support](https://github.com/3dcitydb/importer-exporter/tree/glb-draco-support).

File: impexp-client/src/main/java/org/citydb/modules/kml/gui/preferences/GeneralPanel.java [hunk 8]
```diff
@@ -338,6 +361,8 @@ public void doTranslation() {
 		gltfConverterBrowseButton.setText(Language.I18N.getString("common.button.browse"));
 		notCreateColladaCheckbox.setText(Language.I18N.getString("pref.kmlexport.label.notCreateColladaFiles"));
 		embedTexturesInGltfCheckbox.setText(Language.I18N.getString("pref.kmlexport.label.embedTexturesInGltfFiles"));
+		exportGltfBinary.setText(Language.I18N.getString("pref.kmlexport.label.exportGltfBinary"));
+		enableGltfDracoCompression.setText(Language.I18N.getString("pref.kmlexport.label.enableGltfDracoCompression"));
```
Comment (Son-HNguyen) [inline impexp-client/src/main/java/org/citydb/modules/kml/gui/preferences/GeneralPanel.java:365]: Moved to branch [glb-draco-support](https://github.com/3dcitydb/importer-exporter/tree/glb-draco-support).

File: impexp-client/src/main/java/org/citydb/modules/kml/gui/preferences/GeneralPanel.java [hunk 9]
```diff
@@ -362,6 +387,8 @@ public void loadSettings() {
 		gltfConverterBrowseText.setText(kmlExporter.getPathOfGltfConverter());
 		notCreateColladaCheckbox.setSelected(kmlExporter.isNotCreateColladaFiles());
 		embedTexturesInGltfCheckbox.setSelected(kmlExporter.isEmbedTexturesInGltfFiles());
+		exportGltfBinary.setSelected(kmlExporter.isExportGltfBinary());
+		enableGltfDracoCompression.setSelected(kmlExporter.isEnableGltfDracoCompression());
```
Comment (Son-HNguyen) [inline impexp-client/src/main/java/org/citydb/modules/kml/gui/preferences/GeneralPanel.java:391]: Moved to branch [glb-draco-support](https://github.com/3dcitydb/importer-exporter/tree/glb-draco-support).

File: impexp-client/src/main/java/org/citydb/modules/kml/gui/preferences/GeneralPanel.java [hunk 10]
```diff
@@ -406,6 +433,8 @@ public void setSettings() {
 		kmlExporter.setPathOfGltfConverter(gltfConverterBrowseText.getText());
 		kmlExporter.setNotCreateColladaFiles(notCreateColladaCheckbox.isSelected());
 		kmlExporter.setEmbedTexturesInGltfFiles(embedTexturesInGltfCheckbox.isSelected());
+		kmlExporter.setExportGltfBinary(exportGltfBinary.isSelected());
+		kmlExporter.setEnableGltfDracoCompression(enableGltfDracoCompression.isSelected());
```
Comment (Son-HNguyen) [inline impexp-client/src/main/java/org/citydb/modules/kml/gui/preferences/GeneralPanel.java:437]: Moved to branch [glb-draco-support](https://github.com/3dcitydb/importer-exporter/tree/glb-draco-support).

File: impexp-client/src/main/java/org/citydb/modules/kml/gui/preferences/GeneralPanel.java [hunk 11]
```diff
@@ -429,7 +458,10 @@ private void setEnabledComponents() {
 		gltfConverterBrowseText.setEnabled(createGltfCheckbox.isSelected());
 		gltfConverterBrowseButton.setEnabled(createGltfCheckbox.isSelected());			
 		notCreateColladaCheckbox.setEnabled(createGltfCheckbox.isSelected());
-		embedTexturesInGltfCheckbox.setEnabled(createGltfCheckbox.isSelected());
+		// if glb is enabled, always embed textures
+		embedTexturesInGltfCheckbox.setEnabled(createGltfCheckbox.isSelected() && !exportGltfBinary.isSelected());
+		exportGltfBinary.setEnabled(createGltfCheckbox.isSelected());
+		enableGltfDracoCompression.setEnabled(createGltfCheckbox.isSelected());
```
Comment (Son-HNguyen) [inline impexp-client/src/main/java/org/citydb/modules/kml/gui/preferences/GeneralPanel.java:464]: Moved to branch [glb-draco-support](https://github.com/3dcitydb/importer-exporter/tree/glb-draco-support).

File: impexp-config/src/main/java/org/citydb/config/project/kmlExporter/KmlExporter.java [hunk 12]
```diff
@@ -94,6 +93,9 @@
 		"pathOfGltfConverter",
 		"notCreateColladaFiles",
 		"embedTexturesInGltfFiles",
+		"exportGltfBinary",
+		"enableGltfDracoCompression",
+		"exportGltfV1",
```
Comment (Son-HNguyen) [inline impexp-config/src/main/java/org/citydb/config/project/kmlExporter/KmlExporter.java:98]: Moved to branch [glb-draco-support](https://github.com/3dcitydb/importer-exporter/tree/glb-draco-support).

File: impexp-client/src/main/java/org/citydb/modules/kml/gui/preferences/GeneralPanel.java (hunk 1 above)
Comment (Son-HNguyen) [inline impexp-client/src/main/java/org/citydb/modules/kml/gui/preferences/GeneralPanel.java:462]: Was NOT moved to branch [glb-draco-support](https://github.com/3dcitydb/importer-exporter/tree/glb-draco-support).

File: impexp-client/src/main/java/org/citydb/modules/kml/gui/preferences/GeneralPanel.java (hunk 2 above)
Comment (Son-HNguyen) [inline impexp-client/src/main/java/org/citydb/modules/kml/gui/preferences/GeneralPanel.java:342]: Was NOT moved to branch [glb-draco-support](https://github.com/3dcitydb/importer-exporter/tree/glb-draco-support).

File: impexp-config/src/main/java/org/citydb/config/project/kmlExporter/KmlExporter.java [hunk 13]
```diff
@@ -178,8 +180,10 @@
 	private boolean createGltfModel;
 	private String pathOfGltfConverter;
 	private boolean notCreateColladaFiles;
-	private boolean embedTexturesInGltfFiles;
 	private boolean exportAsKmz;
+	private boolean embedTexturesInGltfFiles;
+	private boolean exportGltfBinary;
+	private boolean enableGltfDracoCompression;
```
Comment (Son-HNguyen) [inline impexp-config/src/main/java/org/citydb/config/project/kmlExporter/KmlExporter.java:186]: Moved to branch [glb-draco-support](https://github.com/3dcitydb/importer-exporter/tree/glb-draco-support).

File: impexp-config/src/main/java/org/citydb/config/project/kmlExporter/KmlExporter.java [hunk 14]
```diff
@@ -247,7 +251,10 @@ public KmlExporter() {
 		callbackNameJSONP = "handle_3DCityDB_data";
 		createGltfModel = false;
 		notCreateColladaFiles = false;
-		embedTexturesInGltfFiles = false;
+		embedTexturesInGltfFiles = true;
+		exportGltfBinary = false;
+		enableGltfDracoCompression = true;
+		exportGltfV1 = false;
```
Comment (Son-HNguyen) [inline impexp-config/src/main/java/org/citydb/config/project/kmlExporter/KmlExporter.java:257]: Moved to branch [glb-draco-support](https://github.com/3dcitydb/importer-exporter/tree/glb-draco-support).

File: impexp-config/src/main/java/org/citydb/config/project/kmlExporter/KmlExporter.java [hunk 15]
```diff
@@ -455,7 +462,23 @@ public void setEmbedTexturesInGltfFiles(boolean embedTexturesInGltfFiles) {
...
+	public boolean isExportGltfBinary() {
+		return this.exportGltfBinary;
+	}
+
+	public void setEnableGltfDracoCompression(boolean enableGltfDracoCompression) {
+		this.enableGltfDracoCompression = enableGltfDracoCompression;
+	}
+
+	public boolean isEnableGltfDracoCompression() {
+		return this.enableGltfDracoCompression;
```
Comment (Son-HNguyen) [inline impexp-config/src/main/java/org/citydb/config/project/kmlExporter/KmlExporter.java:481]: Moved to branch [glb-draco-support](https://github.com/3dcitydb/importer-exporter/tree/glb-draco-support).

File: impexp-config/src/main/resources/org/citydb/config/i18n/language_de.properties [hunk 16]
```diff
@@ -507,7 +507,9 @@ pref.kmlexport.label.deactivateKmz=KMZ deaktivieren
 pref.kmlexport.label.deactivateGlTF=glTF deaktivieren
 pref.kmlexport.label.kmzGltfWarning=Bei Verwendung des glTF-Formats kann nicht das KMZ-Format verwendet werden
 pref.kmlexport.label.notCreateColladaFiles=COLLADA-Dateien (.dae) nicht erzeugen
-pref.kmlexport.label.embedTexturesInGltfFiles=Texturen in glTF-Dateien (.gltf) einbetten
+pref.kmlexport.label.embedTexturesInGltfFiles=Texturen in glTF- bzw. Binary-glTF-Dateien (.glTF bzw. .glb) einbetten
+pref.kmlexport.label.exportGltfBinary=Binary-glTF- (.glb) anstatt glTF-Dateien (.glTF) exportieren
+pref.kmlexport.label.enableGltfDracoCompression=Geometrien mithilfe von Draco-Kompression exportieren
```
Comment (Son-HNguyen) [inline impexp-config/src/main/resources/org/citydb/config/i18n/language_de.properties:512]: Moved to branch [glb-draco-support](https://github.com/3dcitydb/importer-exporter/tree/glb-draco-support).

File: impexp-config/src/main/resources/org/citydb/config/i18n/language_en.properties [hunk 17]
```diff
@@ -507,7 +507,9 @@ pref.kmlexport.label.deactivateKmz=Deactivate KMZ generation
 pref.kmlexport.label.deactivateGlTF=Deactivate glTF generation
 pref.kmlexport.label.kmzGltfWarning=When choosing glTF generation the option "Export in .kmz format" cannot be used
 pref.kmlexport.label.notCreateColladaFiles=Do not create COLLADA (.dae) files
-pref.kmlexport.label.embedTexturesInGltfFiles=Embed textures in glTF (.gltf) files
+pref.kmlexport.label.embedTexturesInGltfFiles=Embed textures in glTF (.gltf) or binary glTF (.glb) files
+pref.kmlexport.label.exportGltfBinary=Export binary glTF (.glb) instead of glTF (.gltf) files
+pref.kmlexport.label.enableGltfDracoCompression=Export using Draco geometry compression
```
Comment (Son-HNguyen) [inline impexp-config/src/main/resources/org/citydb/config/i18n/language_en.properties:512]: Moved to branch [glb-draco-support](https://github.com/3dcitydb/importer-exporter/tree/glb-draco-support).

File: impexp-kml-collada-plugin/src/main/java/org/citydb/modules/kml/database/KmlExporterManager.java [hunk 18]
```diff
@@ -650,23 +663,137 @@ public void print(ColladaBundle colladaBundle, long id, boolean balloonInSeparat
...
-			ProcessBuilder pb = new ProcessBuilder(collada2gltfFile.getAbsolutePath(), "-i", colladaBundle.getGmlId() + ".dae", "-o", gltfModelFile.getAbsolutePath(), "-v", exportGltfV1 ? "1.0" : "2.0");
+			List<String> commands = new ArrayList<>();
+			commands.add(collada2gltfFile.getAbsolutePath());
+			commands.add("-i");
+			commands.add(colladaModelFile.getAbsolutePath());
+			commands.add("-o");
+			commands.add(gltfModelFile.getAbsolutePath());
+			commands.add("-v");
+			commands.add(exportGltfV1 ? "1.0" : "2.0");
+			if (!config.getProject().getKmlExporter().isEmbedTexturesInGltfFiles()) {
```
Comment (Son-HNguyen) [inline impexp-kml-collada-plugin/src/main/java/org/citydb/modules/kml/database/KmlExporterManager.java:698]: Moved to branch [glb-draco-support](https://github.com/3dcitydb/importer-exporter/tree/glb-draco-support).

File: impexp-kml-collada-plugin/src/main/java/org/citydb/modules/kml/database/KmlExporterManager.java [hunk 19]
```diff
@@ -650,23 +663,137 @@ public void print(ColladaBundle colladaBundle, long id, boolean balloonInSeparat
...
+				// always embed textures if glb is enabled, this is handled in the GUI
+				commands.add("-t");
+			}
+			if (config.getProject().getKmlExporter().isExportGltfBinary()) {
+				commands.add("-b");
+			}
+			if (config.getProject().getKmlExporter().isEnableGltfDracoCompression()) {
+				commands.add("-d");
+			}
+			ProcessBuilder pb = new ProcessBuilder(commands);
```
Comment (Son-HNguyen) [inline impexp-kml-collada-plugin/src/main/java/org/citydb/modules/kml/database/KmlExporterManager.java:708]: Moved to branch [glb-draco-support](https://github.com/3dcitydb/importer-exporter/tree/glb-draco-support).

File: impexp-kml-collada-plugin/src/main/java/org/citydb/modules/kml/database/KmlExporterManager.java [hunk 20]
```diff
@@ -650,23 +663,137 @@ public void print(ColladaBundle colladaBundle, long id, boolean balloonInSeparat
...
+			List<String> commands = new ArrayList<>();
+			commands.add(collada2gltfFile.getAbsolutePath());
+			commands.add("-i");
+			commands.add(colladaModelFile.getAbsolutePath());
+			commands.add("-o");
+			commands.add(gltfModelFile.getAbsolutePath());
+			commands.add("-v");
+			commands.add(exportGltfV1 ? "1.0" : "2.0");
+			if (!config.getProject().getKmlExporter().isEmbedTexturesInGltfFiles()) {
+				// always embed textures if glb is enabled, this is handled in the GUI
```
Comment (Son-HNguyen) [inline impexp-kml-collada-plugin/src/main/java/org/citydb/modules/kml/database/KmlExporterManager.java:699]: Was NOT moved to branch [glb-draco-support](https://github.com/3dcitydb/importer-exporter/tree/glb-draco-support).

File: impexp-kml-collada-plugin/src/main/java/org/citydb/modules/kml/database/KmlExporterManager.java [hunk 21]
```diff
@@ -650,23 +663,137 @@ public void print(ColladaBundle colladaBundle, long id, boolean balloonInSeparat
...
+					Writer writer = new BufferedWriter(new FileWriter(gltfModelFile.getAbsolutePath()));
+					writer.write(sb.toString());
+					writer.close();
+
+					// remove the image file
+					(new File(buildingDirectory + File.separator + texImageUri)).delete();
+				}
 			} catch (IOException | InterruptedException e) {
-				log.debug("Unexpected errors occurred while converting collada to glTF for city object '" + colladaBundle.getGmlId() + "' with output path: '" + gltfModelFile.getAbsolutePath() + "'");
+				log.debug("Unexpected errors occurred while converting collada to glTF for city object '" + colladaBundle.getGmlId() + "' with output path: '" + gltfModelFile.getAbsolutePath() + "'" + "\n" + e.getMessage());
```
Comment (Son-HNguyen) [inline impexp-kml-collada-plugin/src/main/java/org/citydb/modules/kml/database/KmlExporterManager.java:756]: Moved to branch [glb-draco-support](https://github.com/3dcitydb/importer-exporter/tree/glb-draco-support).

File: impexp-kml-collada-plugin/src/main/java/org/citydb/modules/kml/database/KmlExporterManager.java [hunk 22]
```diff
@@ -650,23 +663,137 @@ public void print(ColladaBundle colladaBundle, long id, boolean balloonInSeparat
...
 			} finally {
-				if (config.getProject().getKmlExporter().isNotCreateColladaFiles() && gltfModelFile.exists()) {
+				if (config.getProject().getKmlExporter().isNotCreateColladaFiles() && (gltfModelFile.exists() || (new File(gltfModelFile.getAbsolutePath().replace(".gltf", ".glb"))).exists())) {
```
Comment (Son-HNguyen) [inline impexp-kml-collada-plugin/src/main/java/org/citydb/modules/kml/database/KmlExporterManager.java:758]: Moved to branch [glb-draco-support](https://github.com/3dcitydb/importer-exporter/tree/glb-draco-support).

**Comment by Son-HNguyen** [comment]
For the second branch to handle implicit objects, after a discussion with @yaozhihang, we are reconsidering whether:
+ We should store texture images of all objects in one folder `Appearances` and implicit objects (`.dae`, `.gltf` and `.glb`) in the folder `ImplicitObjects` (as implemented so far in this PR). This removes most of the file redundancies but may cause some performance issues when the number of texture files becomes very large (since all textures are stored in one single folder).
+ OR we should store all implicit objects (`.dae`, `.gltf` and `.glb`) and their texture images in one folder `ImplicitObjects`. Other objects that are not implicit shall be exported in the same way as in `master`, i.e. to export them (`.dae`, `.gltf` and `.glb`) together with their texture images in the respective tiles folders. This could improve the visualization performance, since not all textures are stored in the same folder, but might contain some redundancies.

**Comment by clausnagel** [comment]
@Son-HNguyen, @yaozhihang: what is the status of this PR? Are you still working on this?

**Comment by Son-HNguyen** [comment]
This PR was split into two halves, the first half has already been merged to `master`. The second half, which is about restructuring the export directories (see my comment above https://github.com/3dcitydb/importer-exporter/pull/112#issuecomment-697285653), has not been further worked on, since, if I understood him correctly, @yaozhihang prefers the existing structure of the export directories. For archiving purposes, we'd suggest still keeping this PR online (and marking it as a draft as you already did).

**Comment by yaozhihang** [comment]
@Son-HNguyen yes, I preferred the second option, because putting a large number of all textures images into a single folder may cause performance problem in Windows OS. Anyway, the implementation of the 1. or 2. option is quite easy.

The difficult thing is how to handle the archived exports (KMZ). The key question is how to reference the external implicit object (e.g. .dae) from the archived KML file. I think @Son-HNguyen is working hard on this currently.

**Comment by clausnagel** [comment]
Closing this PR as there is no obvious development going on. Please start a new issue or a new PR for parts of this PR if they make sense and can be completed.

**Pull request closed without merging**